lockfile/src/parse_depfile.rs

```diff
@@ -0,0 +1,249 @@
+//! Parse generic dependency files.
+use std::ffi::OsStr;
+#[cfg(feature = "generator")]
+use std::path::Path;
+use std::path::PathBuf;
+
+use anyhow::{anyhow, Context};
+use phylum_types::types::package::{PackageDescriptor, PackageDescriptorAndLockfile};
+use serde::{Deserialize, Serialize};
+
+use crate::{LockfileFormat, Package, PackageVersion, Parse, ThirdPartyVersion};
+
+/// Lockfile parsing error.
+#[derive(thiserror::Error, Debug)]
+pub enum ParseError {
+    /// Dependency file is a manifest, but lockfile generation is disabled.
+    #[error("Parsing {0:?} requires lockfile generation, but it was disabled")]
+    ManifestWithoutGeneration(String),
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+/// Lockfile with all its dependencies.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ParsedLockfile {
+    /// User-facing lockfile identifier (i.e. a file path or name).
+    pub path: String,
+    pub packages: Vec<PackageDescriptor>,
+    pub format: LockfileFormat,
+}
+
+impl ParsedLockfile {
+    pub fn new(
+        path: impl Into<String>,
+        format: LockfileFormat,
+        packages: Vec<PackageDescriptor>,
+    ) -> Self {
+        Self { path: path.into(), packages, format }
+    }
+
+    /// Convert packages to API's expected format.
+    pub fn api_packages(&self) -> Vec<PackageDescriptorAndLockfile> {
+        self.packages
+            .iter()
+            .map(|package_descriptor| PackageDescriptorAndLockfile {
+                package_descriptor: package_descriptor.clone(),
+                lockfile: Some(self.path.clone()),
+            })
+            .collect()
+    }
+}
+
+/// Parse a dependency file.
+///
+/// The `path` argument is only used for display purposes.
+///
+/// The `_generation_path` must point to the manifest on the filesystem if
+/// lockfile generation should be performed. Use `None` to disable lockfile
+/// generation.
+///
+/// # Features
+///
+/// Parsing manifests requires the `generator` feature.
+pub fn parse_depfile(
+    contents: &str,
+    file_name: Option<&OsStr>,
+    format: Option<LockfileFormat>,
+    path: &str,
+    _generation_path: Option<PathBuf>,
```

Comment:
Can't this just be a `bool`?

```suggestion
    _generate_lockfiles: bool,
```

If set to `true`, it can simply use the path from the `file_name` parameter.

For example, the current caller [here][1]:

[1]: https://github.com/phylum-dev/cli/blob/37a219e5218b049a3f1be3db5a242427a6783de1/cli/src/commands/parse.rs#L183-L186

```rs
let generation_path = generate_lockfiles.then(|| path.clone());
let file_name = path.file_name();

phylum_lockfile::parse_depfile(&contents, file_name, format, &id, generation_path)
```

Could simply be this:

```rs
phylum_lockfile::parse_depfile(&contents, Some(path), format, &id, generate_lockfiles)
```

Reply:
No, because `file_name` is a file name. Not a path.

Reply:
That's an arbitrary restriction though, right? The code would still function if a full path was passed for that argument.

Reply:
We need two different things here:
 - A file name to determine lockfile/manifest format and maybe_manifest/lockfile without filesystem access
 - A path that is used for lockfile generation

The real path is only needed for lockfile generation, so using it as an option makes sense to me.

This change is mainly due to how an API is interacted with. Say you upload a file, then the full path even if accessible isn't really useful. But the file name can still be passed to the API to help with lockfile identification.